ably/realtime_conn_spec_test.go

```diff
@@ -1078,3 +1078,64 @@ func TestRealtimeConn_RTN16(t *testing.T) {
 func sameConnection(a, b string) bool {
 	return strings.Split(a, "-")[0] == strings.Split(b, "-")[0]
 }
+
+type protoConnWithReceiveHook struct {
+	proto.Conn
+	ok chan *proto.ProtocolMessage
+}
+
+func (w *protoConnWithReceiveHook) Receive(deadline time.Time) (*proto.ProtocolMessage, error) {
+	msg, err := w.Conn.Receive(deadline)
+	w.ok <- msg
+	return msg, err
+}
+
+func TestRealtimeConn_RTN23(t *testing.T) {
+	t.Parallel()
+	var query url.Values
+	ok := make(chan *proto.ProtocolMessage, 2)
+	timeout := 30 * time.Millisecond
+	app, c := ablytest.NewRealtime(&ably.ClientOptions{
+		RealtimeRequestTimeout: timeout,
+		Dial: func(protocol string, u *url.URL) (proto.Conn, error) {
+			query = u.Query()
+			conn, err := ablyutil.DialWebsocket(protocol, u)
+			if err != nil {
+				return nil, err
+			}
+			return &protoConnWithReceiveHook{ok: ok, Conn: conn}, nil
+		},
+	})
+	defer safeclose(t, ablytest.FullRealtimeCloser(c), app)
+
+	err := ablytest.ConnWaiter(c, c.Connect, ably.ConnectionEventConnected).Wait()
+	if err != nil {
+		t.Fatal(err)
+	}
+	msg := <-ok // consume the connect message
+	maxIdleInterval := time.Duration(msg.ConnectionDetails.MaxIdleInterval) * time.Millisecond
+	receiveTimeout := timeout + maxIdleInterval
+
+	{ // RTN23b
+		h := query.Get("heartbeats")
+		if h != "true" {
+			t.Errorf("expected heartbeats query param to be true got %q", h)
+		}
+	}
+	{ //RTN23a
+		select {
+		case msg := <-ok:
+			// The test here is ensuring we receive a heartbeat within the timeout bounds
+			// as per rtn23a because the connection is idle since receiving connection
+			// message from ably.
+			if msg == nil {
+				t.Fatal("expected to receive heartbeat message")
+			}
+			if msg.Action != proto.ActionHeartbeat {
+				t.Errorf("expected action %v got %v", proto.ActionHeartbeat, msg.Action)
+			}
+		case <-time.After(receiveTimeout):
```

Comment:
Sorry for being late here, but a few comments.

This test is very flaky and fails a lot on my machine with:

```
$ go test -race -count=1 -v -run='TestRealtimeConn_RTN23' ./...
=== RUN   TestRealtimeConn_RTN23
=== PAUSE TestRealtimeConn_RTN23
=== CONT  TestRealtimeConn_RTN23
    TestRealtimeConn_RTN23: realtime_conn_spec_test.go:1146: no heartbeat was sent by ably
```

Also, we're missing a case here. [RTN23a says](https://docs.ably.io/client-lib-development-guide/features/#RTN23a):

> If a transport does not receive any indication of activity on a transport for a period greater than the sum of the maxIdleInterval (which will be sent in the connectionDetails of the most recent CONNECTED message received on that transport) and the realtimeRequestTimeout, that transport should be disconnected. 

But we're not testing this case. We're testing the case in which Ably _does_ send a heartbeat in time. We're missing a test that checks that we disconnect if we don't get a heartbeat.

Also, note that `maxIdleInterval` is quite big (15 s). I think we should mock a shorter `maxIdleInterval`, so that tests don't take too long. (We also avoid flakiness this way.)

Reply:
PS. I said "on my machine" but it's failing a lot in CI too.

Reply:
@tcard its weird, that test is passing , it takes a long time though

```
$ go test -v -run TestRealtimeConn_RTN23 ./ably/
=== RUN   TestRealtimeConn_RTN23
=== PAUSE TestRealtimeConn_RTN23
=== CONT  TestRealtimeConn_RTN23
--- PASS: TestRealtimeConn_RTN23 (25.46s)
PASS
ok      github.com/ably/ably-go/ably    25.477s
```

Reply:
See a [recent occurrence](https://travis-ci.org/github/ably/ably-go/jobs/713709989) of this test failing, and [another](https://travis-ci.org/github/ably/ably-go/jobs/713709990). The second one with a different message:

```
--- FAIL: TestRealtimeConn_RTN23 (37.50s)
    realtime_conn_spec_test.go:1140: expected to receive heartbeat message
```

Also, what about this?

> Also, we're missing a case here. [RTN23a says](https://docs.ably.io/client-lib-development-guide/features/#RTN23a):
> 
> > If a transport does not receive any indication of activity on a transport for a period greater than the sum of the maxIdleInterval (which will be sent in the connectionDetails of the most recent CONNECTED message received on that transport) and the realtimeRequestTimeout, that transport should be disconnected.
> 
> But we're not testing this case. We're testing the case in which Ably _does_ send a heartbeat in time. We're missing a test that checks that we disconnect if we don't get a heartbeat.

Reply:
I've disabled the test at #178 to keep CI stable. Should be enabled again once this is addresses.